src/main/resources/com/rtbhouse/model/natives/FactorizationMachinesNativeOps.h

```diff
@@ -0,0 +1,173 @@
+#include <cstdlib>
+#include <cmath>
+#include <inttypes.h>
+#include <pmmintrin.h>
+#include <immintrin.h>
+
+static uint32_t const FACTOR_SIZE = 4;
+
+// Horizontal sum of 8 floats
+inline float horizontal_add8(__m256 x) {
+    __m128 const hi256 = _mm256_extractf128_ps(x, 1);
+    __m128 const lo256 = _mm256_castps256_ps128(x);
+    __m128 const sum2 = _mm_add_ps(lo256, hi256); // x3 + x7, x2 + x6, x1 + x5, x0 + x4
+
+    __m128 const lo128 = sum2;
+    __m128 const hi128 = _mm_movehl_ps(sum2, sum2);
+    __m128 const sum4 = _mm_add_ps(lo128, hi128); //  _, _, (x3 + x7) + (x1 + x5), (x0 + x4) + (x2 + x6)
+
+    __m128 const lo = sum4;
+    __m128 const hi = _mm_shuffle_ps(sum4, sum4, 0x1);
+    __m128 const sum8 = _mm_add_ss(lo, hi); // _, _, _, [ (x3 + x7) + (x1 + x5) ]  +  [ (x0 + x4) + (x2 + x6) ]
+
+    return _mm_cvtss_f32(sum8);
+}
+
+// Horizontal sum of 4 floats
+inline float horizontal_add4(__m128 x) {
+    __m128 const sum2 = _mm_hadd_ps(x, x); // x3 + x2, x1 + x0, _, _
+    __m128 const sum4 = _mm_hadd_ps(sum2, sum2); // x3 + x2  +  x1 + x0, _, _
+
+    return _mm_cvtss_f32(sum4);
+}
+
+// Performs:
+//
+//  [ sum ] += [ w1 ] * [ w2lo, w2hi ]
+//
+// in a vectorized manner.
+//
+// Where:
+//   - w1 points to 8 floats
+//   - w2lo points to 4 floats
+//   - w2hi points to 4 floats
+inline void mul_add8(const float * const w1, const float * const w2lo, const float * const w2hi, __m256 & sum) {
+    __m256 YMMw1 = _mm256_loadu_ps(w1);
+    __m256 YMMw2 = _mm256_setzero_ps();
+
+    __m128 const XMMw2lo = _mm_load_ps(w2lo);
+    __m128 const XMMw2hi = _mm_load_ps(w2hi);
+    YMMw2 = _mm256_insertf128_ps(YMMw2, XMMw2lo, 0);
+    YMMw2 = _mm256_insertf128_ps(YMMw2, XMMw2hi, 1);
+
+    // GCC will translate it to fused multiply–add (_mm256_fmadd_ps) anyway if -mfma switch is enabled
+    sum = _mm256_add_ps(sum, _mm256_mul_ps(YMMw1, YMMw2));
+}
+
+// Performs:
+//
+//   [ sum ] += [ w1 ] * [ w2 ]
+//
+// in a vectorized manner.
+//
+// Where:
+//   - w1 points to 4 floats
+//   - w2 points to 4 floats
+inline void mul_add4(const float * const w1, const float * const w2, __m128 & sum) {
+    __m128 const XMMw1 = _mm_load_ps(w1);
+    __m128 const XMMw2 = _mm_load_ps(w2);
+    sum = _mm_add_ps(sum, _mm_mul_ps(XMMw1, XMMw2));
+}
+
+// Performs:
+//
+//  [ sum ] += dot(weights[feature1, field2]), weights[feature2, field1])
```

Comment:
redundant right parentases `(` in the middle of `dot()`
same issue in latter function doc